bin.src/generateCommands.py

```diff
@@ -0,0 +1,581 @@
+                then
+                    mkdir -p {shlex.quote(fiberTraceDir)}/COMBINED
+                    combineFiberTraces.py \
+                        {shlex.quote(fiberTraceDir)}/COMBINED/"$(basename "$trace0")" \
+                        $(find {shlex.quote(fiberTraceDir)} -maxdepth 1 -name 'pfsFiberTrace-*-{detector}.fits' -print) || return 1
+                fi
+            """)
+            print(command[1:].rstrip())
+
+        ingestCalibs("FIBERTRACE/COMBINED", dataDir, calib, rerun, copyMode)
+        #ingestCalibs("DETECTORMAP", dataDir, calib, rerun, copyMode, overwrite=True)
+
+
+@CalibSource.registerAsSubclass("arc")
+class ArcSource(CalibSource):
+    commandName = "reduceArc.py"
+
+    def execute(self, dataDir, calib, rerun, copyMode, processes=1, verbose=False):
+        visits = visitsToString(self.visits)
+
+        command = [
+            self.commandName,
+            dataDir,
+            f"--calib={os.path.join(dataDir, calib)}",
+            f"--rerun={rerun}",
+            "--id", f"visit={visits}",
+            "-j", processes,
+            ]
+
+        if self.configs:
+            command += ["--config"] + self.configs
+
+        print(f"{shellCommand(command)} || return 1")
+
+        #ingestCalibs(self.typeName, dataDir, calib, rerun, copyMode)
+        ingestCalibs("DETECTORMAP", dataDir, calib, rerun, copyMode, overwrite=True)
+
+
+class CalibBlock:
+    calibTypes = [] # will be replaced in processYaml()
+
+    def __init__(self, blockName:str, yamlBlock:Mapping[str,Any]):
+        self.name = blockName
+        self.sources = {}
+
+        for dt, block in yamlBlock.items():
+            if dt not in self.calibTypes:
+                print("Saw unknown calibType \"%s\" in block %s (expected %s)" %
+                      (dt, blockName, ", ".join(self.calibTypes)))
+                continue
+
+            self.sources[dt] = CalibSource.getSubclass(dt).fromYaml(block)
+
+    def execute(self, dataDir, calib, rerun, copyMode, calibTypes=None, processes=1, verbose=False):
+        """Execute a calibration block"""
+        if calibTypes is None:
+            calibTypes = self.calibTypes
+
+        if verbose:
+            print(f"Processing calib block '{self.name}'", file=sys.stderr)
+        for dt in calibTypes:
+            if dt in self.sources:
+                self.sources[dt].execute(dataDir, calib, f"{rerun}/{dt}", copyMode, processes=processes, verbose=verbose)
+
+
+class ScienceStep:
+    __slots__ = ["configs"]
+    __subclasses:Dict[str,type] = {}
+
+    def __init__(self, configs:Optional[Iterator[str]]):
+        self.configs = list(configs) if configs else []
+
+    @classmethod
+    def fromYaml(cls, yamlBlock:Mapping[str,Any]) -> "ScienceStep":
+        r"""
+        \param yamlBlock: {
+            "config": ...,
+            "configs": ...,
+            }
+        """
+        configs = _getConfigs(yamlBlock)
+        return cls(configs)
+
+    @staticmethod
```

Comment:
Surely this is a `classmethod`? But I'm hoping it will go away when you use `__subclasses__`.

Reply:
This has to be `classmethod` because `Subclass.fromYaml()` has to return `Subclass` instead of `ScienceStep`. The `cls` argument is essential.